Add CJsonParser, revise io_read_all/io_read_all_str

Robyt3 opened the pull request:
#### `io_read_all`

- Allocate necessary memory for the entire file at once by first getting its length, instead of reallocating memory multiple times while reading the file.
- Do not zero-terminate the loaded data, as `io_read_all_str` is supposed to be used for zero-terminated strings instead.
- Return pointer to result or 0 on error for more convenient usage.

#### `io_read_all_str`

- Allocate necessary memory for the entire file at once by first getting its length, instead of reallocating memory multiple times while reading the file.
- As before: zero-terminate the loaded data and check for null-bytes inside the string.

#### `CStorage::ReadFile`

- Return true instead of false on success for consistency with other C++ style methods that return truthy results on success. The method was previously unused so no usage needs to be adjusted.
- Use the method for loading font file into memory.

### `CJsonParser`

- `CJsonParser` wraps the file and error handling when using the json parser.
- Methods to parse json data from from memory are also added to simplify testing the class.
- The loaded json data is automatically freed when the `CJsonParser` instance is destructed.
- All usages of the json parser are refactored using this class, removing a lot of duplicate code.
- The json library must now also be linked in the server, tools, mastersrv and versionsrv build with bam. It was already linked with those targets with cmake.

heinrich5991 commented:
> * Return pointer to result or 0 on error for more convenient usage.

I specifically avoided this because you cannot use the pointer safely without also checking the returned length. I don't really see a use case where you can use the returned pointer in a way that is safe, and cannot also be done in the same length with the out parameter.

However, this obviously doesn't mean that you're not free to change it. ^^

heinrich5991 commented:
> * Return pointer to result or 0 on error for more convenient usage.

I specifically avoided this because you cannot use the pointer safely without also checking the returned length. I don't really see a use case where you can use the returned pointer in a way that is safe, and cannot also be done in the same length with the out parameter.

However, this obviously doesn't mean that you're not free to change it. ^^

heinrich5991 commented:
The new `io_read_all` and `io_read_all_str` have a TOCTOU (time-of-check vs time-of-use) problem. The file size might have changed by the time the file is read. This means it might return incomplete data. I think the only way to correct this is to reintroduce the loop and keep reading if the file is larger than the advertised size.

Robyt3 commented:
> I specifically avoided this because you cannot use the pointer safely without also checking the returned length. I don't really see a use case where you can use the returned pointer in a way that is safe, and cannot also be done in the same length with the out parameter.

After looking at it again, I agree, so I changed it back again.

> The new `io_read_all` and `io_read_all_str` have a TOCTOU (time-of-check vs time-of-use) problem. The file size might have changed by the time the file is read. This means it might return incomplete data. I think the only way to correct this is to reintroduce the loop and keep reading if the file is larger than the advertised size.

I rewrote `io_read_all` so it should only need one allocation per file in cases where the actual file size matches the expected file size. Otherwise it falls back to doubling the buffer size if the actual file size is larger than expected. If the actual size is smaller, the len is also correctly adjusted now, which was not the case with the previous implementation.

heinrich5991 commented on src/base/system.c:
```diff
@@ -381,28 +381,33 @@ unsigned io_read(IOHANDLE io, void *buffer, unsigned size)
+				buffer = realloc(buffer, cap);
+			}
+		}
+		if(len <= cap)
```
It looks to me as if this condition is always true.

heinrich5991 commented:
Both changes look good, thanks! :)
